crates/wasi/src/preview2/filesystem.rs

```diff
@@ -180,35 +177,103 @@ pub(crate) enum FileOutputMode {
+
+#[async_trait::async_trait]
+impl HostOutputStream for FileOutputStream {
+    fn write(&mut self, buf: Bytes) -> Result<(), OutputStreamError> {
         use system_interface::fs::FileIoExt;
+
+        if self.closed {
+            return Err(OutputStreamError::Closed);
+        }
+        if !matches!(self.task, MaybeDone::Gone) {
+            // a write is pending - this call was not permitted
+            return Err(OutputStreamError::Trap(anyhow!(
+                "write not permitted: FileOutputStream write pending"
+            )));
+        }
         let f = Arc::clone(&self.file);
         let m = self.mode;
-        let r = tokio::task::spawn_blocking(move || match m {
-            FileOutputMode::Position(p) => f.write_at(buf.as_ref(), p),
-            FileOutputMode::Append => f.append(buf.as_ref()),
-        })
-        .await
-        .unwrap();
-        let (n, state) = write_result(r)?;
-        if let FileOutputMode::Position(ref mut position) = self.mode {
-            *position += n as u64;
+        self.task = maybe_done(AbortOnDropJoinHandle::from(tokio::task::spawn_blocking(
+            move || match m {
+                FileOutputMode::Position(mut p) => {
+                    let mut buf = buf;
+                    while !buf.is_empty() {
+                        let nwritten = f.write_at(buf.as_ref(), p)?;
+                        // afterwards buf contains [nwritten, len):
+                        let _ = buf.split_to(nwritten);
+                        p += nwritten as u64;
+                    }
+                    Ok(())
+                }
+                FileOutputMode::Append => {
+                    let mut buf = buf;
+                    while !buf.is_empty() {
+                        let nwritten = f.append(buf.as_ref())?;
+                        let _ = buf.split_to(nwritten);
+                    }
+                    Ok(())
+                }
+            },
+        )));
+        Ok(())
+    }
+    fn flush(&mut self) -> Result<(), OutputStreamError> {
+        if self.closed {
+            return Err(OutputStreamError::Closed);
+        }
+        // Only userland buffering of file writes is in the blocking task.
+        Ok(())
+    }
+    async fn write_ready(&mut self) -> Result<usize, OutputStreamError> {
+        if self.closed {
+            return Err(OutputStreamError::Closed);
+        }
+        // If there is no outstanding task, accept more input:
+        if matches!(self.task, MaybeDone::Gone) {
+            return Ok(FILE_WRITE_CAPACITY);
+        }
+        // Wait for outstanding task:
+        std::pin::Pin::new(&mut self.task).await;
+
+        // Mark task as finished, and handle output:
+        match std::pin::Pin::new(&mut self.task)
+            .take_output()
+            .expect("just awaited for MaybeDone completion")
+            .context("join of FileOutputStream worker task")
+            .map_err(OutputStreamError::Trap)?
```

Comment:
This feels a little odd to me in the sense that the trap/context make sense if this is going to be a trap/complete failure but this is also being wrapped up in `LastOperationFailed` below additionally. My guess, having not read further yet, is that there's error handling of `OutputStreamError` which makes this not always a trap. This looks like, given the wrapper on this line though, that it's always a trap?

I'll dig in more when seeing the error handling.

Reply:
Oh actually I think I'm totally wrong. I think this is handling the `Result` layer of the spawned task which is propagating panics.

In that case, could `unwrap` or `expect` be used here instead of `context`/`map_err`?

Reply:
yes, this should be an expect instead of a trap.